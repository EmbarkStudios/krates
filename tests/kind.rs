mod util;

use util::build;

#[test]
fn all_the_things() {
    let grafs = build("all-features.json", krates::Builder::new()).unwrap();

    assert!(grafs.filtered.is_empty());
    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn filters_dev() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();

        // This shouldn't actually affect anything, as dev dependencies
        // for non-workspace crates are already not resolved
        assert!(grafs.filtered.is_empty());
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::All);

        // This will be equivalent to to filtering workspace dev crates
        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn filters_build() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn filters_normal() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Normal, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Normal, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Normal, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn filters_build_and_dev() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::NonWorkspace);
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::Workspace);
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::All);
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn only_b() {
    let mut kb = krates::Builder::new();
    kb.include_workspace_crates(["/home/jake/code/krates/tests/ws/b/Cargo.toml"]);

    let grafs = build("all-features.json", kb).unwrap();
    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn filters_after_build() {
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();

        let filtered = grafs.actual.krates_filtered(krates::DepKind::Dev);

        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();

        let expected = format!("{:#?}", grafs.actual.krates().collect::<Vec<_>>());
        let actual = format!("{filtered:#?}");

        similar_asserts::assert_eq!(expected, actual);
    }

    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();

        let filtered = grafs.actual.krates_filtered(krates::DepKind::Build);

        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();

        let expected = format!("{:#?}", grafs.actual.krates().collect::<Vec<_>>());
        let actual = format!("{filtered:#?}");

        similar_asserts::assert_eq!(expected, actual);
    }
}
