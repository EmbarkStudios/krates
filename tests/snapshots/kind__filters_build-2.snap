---
source: tests/kind.rs
expression: grafs.dotgraph()
---
digraph {
    0 [ label = "crate a 0.1.0 path+file:///krates/tests/ws/a" ]
    1 [ label = "crate b 0.1.0 path+file:///krates/tests/ws/b" ]
    2 [ label = "crate bindgen 0.59.2" ]
    3 [ label = "crate bitflags 1.3.2" ]
    4 [ label = "crate bumpalo 3.11.0" ]
    5 [ label = "crate c 0.1.0 path+file:///krates/tests/ws/c" ]
    6 [ label = "crate cc 1.0.73" ]
    7 [ label = "crate cexpr 0.6.0" ]
    8 [ label = "crate cfg-if 0.1.10" ]
    9 [ label = "crate cfg-if 1.0.0" ]
    10 [ label = "crate clang-sys 1.4.0" ]
    11 [ label = "crate coreaudio-rs 0.9.1" ]
    12 [ label = "crate coreaudio-sys 0.2.10" ]
    13 [ label = "crate difference 2.0.0" ]
    14 [ label = "crate glob 0.3.0" ]
    15 [ label = "crate js-sys 0.3.60" ]
    16 [ label = "crate lazy_static 1.4.0" ]
    17 [ label = "crate lazycell 1.3.0" ]
    18 [ label = "crate leftpad 0.2.0" ]
    19 [ label = "crate libc 0.2.133" ]
    20 [ label = "crate libloading 0.7.3" ]
    21 [ label = "crate log 0.4.17" ]
    22 [ label = "crate memchr 2.5.0" ]
    23 [ label = "crate minimal-lexical 0.2.1" ]
    24 [ label = "crate nix 0.16.1" ]
    25 [ label = "crate nom 7.1.1" ]
    26 [ label = "crate once_cell 1.15.0" ]
    27 [ label = "crate peeking_take_while 0.1.2" ]
    28 [ label = "crate proc-macro2 1.0.44" ]
    29 [ label = "crate quote 1.0.21" ]
    30 [ label = "crate regex 1.6.0" ]
    31 [ label = "crate regex-syntax 0.6.27" ]
    32 [ label = "crate ring 0.16.20" ]
    33 [ label = "crate rustc-hash 1.1.0" ]
    34 [ label = "crate shlex 1.1.0" ]
    35 [ label = "crate spin 0.5.2" ]
    36 [ label = "crate syn 1.0.101" ]
    37 [ label = "crate unicode-ident 1.0.4" ]
    38 [ label = "crate untrusted 0.7.1" ]
    39 [ label = "crate void 1.0.2" ]
    40 [ label = "crate wasm-bindgen 0.2.83" ]
    41 [ label = "crate wasm-bindgen-backend 0.2.83" ]
    42 [ label = "crate wasm-bindgen-futures 0.4.33" ]
    43 [ label = "crate wasm-bindgen-macro 0.2.83" ]
    44 [ label = "crate wasm-bindgen-macro-support 0.2.83" ]
    45 [ label = "crate wasm-bindgen-shared 0.2.83" ]
    46 [ label = "crate web-sys 0.3.60" ]
    47 [ label = "crate winapi 0.2.8" ]
    48 [ label = "crate winapi 0.3.9" ]
    49 [ label = "crate winapi-i686-pc-windows-gnu 0.4.0" ]
    50 [ label = "crate winapi-x86_64-pc-windows-gnu 0.4.0" ]
    51 [ label = "feature default" ]
    52 [ label = "feature default" ]
    53 [ label = "feature default" ]
    54 [ label = "feature clang_6_0" ]
    55 [ label = "feature runtime" ]
    56 [ label = "feature std" ]
    57 [ label = "feature unicode" ]
    58 [ label = "feature default" ]
    59 [ label = "feature default" ]
    60 [ label = "feature default" ]
    61 [ label = "feature default" ]
    62 [ label = "feature Crypto" ]
    63 [ label = "feature Window" ]
    64 [ label = "feature std" ]
    65 [ label = "feature audio_toolbox" ]
    66 [ label = "feature audio_unit" ]
    67 [ label = "feature core_audio" ]
    68 [ label = "feature core_midi" ]
    69 [ label = "feature open_al" ]
    70 [ label = "feature runtime" ]
    71 [ label = "feature default" ]
    72 [ label = "feature errhandlingapi" ]
    73 [ label = "feature libloaderapi" ]
    74 [ label = "feature extra_traits" ]
    75 [ label = "feature default" ]
    76 [ label = "feature default" ]
    77 [ label = "feature std" ]
    78 [ label = "feature std" ]
    79 [ label = "feature proc-macro" ]
    80 [ label = "feature unicode" ]
    81 [ label = "feature unicode-age" ]
    82 [ label = "feature unicode-bool" ]
    83 [ label = "feature unicode-case" ]
    84 [ label = "feature unicode-gencat" ]
    85 [ label = "feature unicode-perl" ]
    86 [ label = "feature unicode-script" ]
    87 [ label = "feature unicode-segment" ]
    88 [ label = "feature std" ]
    89 [ label = "feature ntsecapi" ]
    90 [ label = "feature wtypesbase" ]
    91 [ label = "feature proc-macro" ]
    92 [ label = "feature spans" ]
    93 [ label = "feature default" ]
    94 [ label = "feature default" ]
    95 [ label = "feature default" ]
    96 [ label = "feature default" ]
    97 [ label = "feature full" ]
    98 [ label = "feature default" ]
    99 [ label = "feature MessageEvent" ]
    100 [ label = "feature Worker" ]
    101 [ label = "feature spans" ]
    102 [ label = "feature visit" ]
    103 [ label = "feature spans" ]
    104 [ label = "feature leftpad" ]
    105 [ label = "feature leftier-strings" ]
    106 [ label = "feature lazy_static" ]
    107 [ label = "feature libloading" ]
    108 [ label = "feature clang_5_0" ]
    109 [ label = "feature clang_4_0" ]
    110 [ label = "feature clang_3_9" ]
    111 [ label = "feature clang_3_8" ]
    112 [ label = "feature clang_3_7" ]
    113 [ label = "feature clang_3_6" ]
    114 [ label = "feature clang_3_5" ]
    115 [ label = "feature open_al" ]
    116 [ label = "feature audio_toolbox" ]
    117 [ label = "feature audio_unit" ]
    118 [ label = "feature core_audio" ]
    119 [ label = "feature core_midi" ]
    120 [ label = "feature std" ]
    121 [ label = "feature alloc" ]
    122 [ label = "feature alloc" ]
    123 [ label = "feature race" ]
    124 [ label = "feature unicode-segment" ]
    125 [ label = "feature unicode-script" ]
    126 [ label = "feature unicode-perl" ]
    127 [ label = "feature unicode-gencat" ]
    128 [ label = "feature unicode-case" ]
    129 [ label = "feature unicode-bool" ]
    130 [ label = "feature unicode-age" ]
    131 [ label = "feature once_cell" ]
    132 [ label = "feature dev_urandom_fallback" ]
    133 [ label = "feature alloc" ]
    134 [ label = "feature std" ]
    135 [ label = "feature std" ]
    136 [ label = "feature quote" ]
    137 [ label = "feature proc-macro" ]
    138 [ label = "feature printing" ]
    139 [ label = "feature parsing" ]
    140 [ label = "feature derive" ]
    141 [ label = "feature clone-impls" ]
    142 [ label = "feature std" ]
    143 [ label = "feature std" ]
    144 [ label = "feature spans" ]
    145 [ label = "feature EventTarget" ]
    146 [ label = "feature Event" ]
    0 -> 1 [ label = "" ]
    0 -> 51 [ label = "(dev)" ]
    1 -> 51 [ label = "" ]
    1 -> 52 [ label = "(dev) 'cfg(target_arch = \"x86_64\")'" ]
    1 -> 42 [ label = " 'cfg(all(target_arch = \"wasm32\", target_feature = \"atomics\"))'" ]
    1 -> 42 [ label = " 'cfg(all(target_vendor = \"xboxone\"))'" ]
    2 -> 53 [ label = "" ]
    2 -> 7 [ label = "" ]
    2 -> 54 [ label = "" ]
    2 -> 55 [ label = "" ]
    2 -> 16 [ label = "" ]
    2 -> 17 [ label = "" ]
    2 -> 27 [ label = "" ]
    2 -> 28 [ label = "" ]
    2 -> 29 [ label = "" ]
    2 -> 56 [ label = "" ]
    2 -> 57 [ label = "" ]
    2 -> 58 [ label = "" ]
    2 -> 59 [ label = "" ]
    5 -> 60 [ label = " 'x86_64-apple-darwin'" ]
    5 -> 61 [ label = "(dev)" ]
    5 -> 16 [ label = " 'cfg(any(target_os = \"android\", target_os = \"freebsd\", target_os = \"linux\", target_os = \"netbsd\", target_os = \"openbsd\", target_os = \"solaris\"))'" ]
    5 -> 18 [ label = "" ]
    5 -> 19 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    5 -> 24 [ label = " 'x86_64-unknown-linux-gnu'" ]
    5 -> 35 [ label = " 'cfg(all(any(target_arch = \"aarch64\", target_arch = \"arm\", target_arch = \"x86\", target_arch = \"x86_64\"), not(target_os = \"ios\")))'" ]
    5 -> 62 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    5 -> 63 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    5 -> 47 [ label = " 'cfg(target_os = \"windows\")'" ]
    7 -> 64 [ label = "" ]
    10 -> 14 [ label = "" ]
    10 -> 14 [ label = "(build)" ]
    10 -> 19 [ label = "" ]
    10 -> 20 [ label = "" ]
    11 -> 53 [ label = "" ]
    11 -> 12 [ label = "" ]
    11 -> 65 [ label = "" ]
    11 -> 66 [ label = "" ]
    11 -> 67 [ label = "" ]
    11 -> 68 [ label = "" ]
    11 -> 69 [ label = "" ]
    12 -> 70 [ label = "(build)" ]
    15 -> 71 [ label = "" ]
    20 -> 9 [ label = " 'cfg(unix)'" ]
    20 -> 72 [ label = " 'cfg(windows)'" ]
    20 -> 73 [ label = " 'cfg(windows)'" ]
    21 -> 9 [ label = "" ]
    24 -> 53 [ label = "" ]
    24 -> 6 [ label = "(build) 'cfg(target_os = \"dragonfly\")'" ]
    24 -> 8 [ label = "" ]
    24 -> 74 [ label = "" ]
    24 -> 75 [ label = "" ]
    24 -> 76 [ label = "" ]
    25 -> 22 [ label = "" ]
    25 -> 77 [ label = "" ]
    25 -> 23 [ label = "" ]
    25 -> 78 [ label = "" ]
    28 -> 37 [ label = "" ]
    29 -> 28 [ label = "" ]
    29 -> 79 [ label = "" ]
    30 -> 31 [ label = "" ]
    30 -> 80 [ label = "" ]
    30 -> 81 [ label = "" ]
    30 -> 82 [ label = "" ]
    30 -> 83 [ label = "" ]
    30 -> 84 [ label = "" ]
    30 -> 85 [ label = "" ]
    30 -> 86 [ label = "" ]
    30 -> 87 [ label = "" ]
    32 -> 6 [ label = "(build)" ]
    32 -> 19 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    32 -> 88 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    32 -> 88 [ label = " 'cfg(any(target_os = \"dragonfly\", target_os = \"freebsd\", target_os = \"illumos\", target_os = \"netbsd\", target_os = \"openbsd\", target_os = \"solaris\"))'" ]
    32 -> 35 [ label = " 'cfg(any(target_arch = \"x86\", target_arch = \"x86_64\", all(any(target_arch = \"aarch64\", target_arch = \"arm\"), any(target_os = \"android\", target_os = \"fuchsia\", target_os = \"linux\"))))'" ]
    32 -> 38 [ label = "" ]
    32 -> 62 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    32 -> 63 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    32 -> 89 [ label = " 'cfg(target_os = \"windows\")'" ]
    32 -> 90 [ label = " 'cfg(target_os = \"windows\")'" ]
    36 -> 28 [ label = "" ]
    36 -> 79 [ label = "" ]
    36 -> 29 [ label = "" ]
    36 -> 91 [ label = "" ]
    36 -> 37 [ label = "" ]
    40 -> 9 [ label = "" ]
    40 -> 43 [ label = "" ]
    40 -> 92 [ label = "" ]
    41 -> 93 [ label = "" ]
    41 -> 21 [ label = "" ]
    41 -> 94 [ label = "" ]
    41 -> 95 [ label = "" ]
    41 -> 96 [ label = "" ]
    41 -> 97 [ label = "" ]
    41 -> 98 [ label = "" ]
    41 -> 45 [ label = "" ]
    42 -> 9 [ label = "" ]
    42 -> 15 [ label = "" ]
    42 -> 71 [ label = "" ]
    42 -> 99 [ label = " 'cfg(target_feature = \"atomics\")'" ]
    42 -> 100 [ label = " 'cfg(target_feature = \"atomics\")'" ]
    43 -> 96 [ label = "" ]
    43 -> 44 [ label = "" ]
    43 -> 101 [ label = "" ]
    44 -> 95 [ label = "" ]
    44 -> 96 [ label = "" ]
    44 -> 102 [ label = "" ]
    44 -> 97 [ label = "" ]
    44 -> 98 [ label = "" ]
    44 -> 41 [ label = "" ]
    44 -> 103 [ label = "" ]
    44 -> 45 [ label = "" ]
    46 -> 15 [ label = "" ]
    46 -> 71 [ label = "" ]
    48 -> 49 [ label = " 'i686-pc-windows-gnu'" ]
    48 -> 50 [ label = " 'x86_64-pc-windows-gnu'" ]
    70 -> 2 [ label = "" ]
    70 -> 55 [ label = "" ]
    53 -> 3 [ label = "" ]
    93 -> 4 [ label = "" ]
    104 -> 5 [ label = "" ]
    104 -> 18 [ label = "" ]
    105 -> 5 [ label = "" ]
    105 -> 104 [ label = "" ]
    106 -> 5 [ label = "" ]
    106 -> 16 [ label = "" ]
    51 -> 5 [ label = "" ]
    51 -> 104 [ label = "" ]
    55 -> 10 [ label = "" ]
    55 -> 107 [ label = "" ]
    107 -> 10 [ label = "" ]
    107 -> 20 [ label = "" ]
    54 -> 10 [ label = "" ]
    54 -> 108 [ label = "" ]
    108 -> 10 [ label = "" ]
    108 -> 109 [ label = "" ]
    109 -> 10 [ label = "" ]
    109 -> 110 [ label = "" ]
    110 -> 10 [ label = "" ]
    110 -> 111 [ label = "" ]
    111 -> 10 [ label = "" ]
    111 -> 112 [ label = "" ]
    112 -> 10 [ label = "" ]
    112 -> 113 [ label = "" ]
    113 -> 10 [ label = "" ]
    113 -> 114 [ label = "" ]
    114 -> 10 [ label = "" ]
    115 -> 11 [ label = "" ]
    115 -> 69 [ label = "" ]
    60 -> 11 [ label = "" ]
    60 -> 116 [ label = "" ]
    60 -> 117 [ label = "" ]
    60 -> 118 [ label = "" ]
    60 -> 115 [ label = "" ]
    60 -> 119 [ label = "" ]
    119 -> 11 [ label = "" ]
    119 -> 68 [ label = "" ]
    118 -> 11 [ label = "" ]
    118 -> 67 [ label = "" ]
    117 -> 11 [ label = "" ]
    117 -> 66 [ label = "" ]
    116 -> 11 [ label = "" ]
    116 -> 65 [ label = "" ]
    69 -> 12 [ label = "" ]
    68 -> 12 [ label = "" ]
    67 -> 12 [ label = "" ]
    66 -> 12 [ label = "" ]
    65 -> 12 [ label = "" ]
    61 -> 13 [ label = "" ]
    120 -> 19 [ label = "" ]
    74 -> 19 [ label = "" ]
    75 -> 19 [ label = "" ]
    75 -> 120 [ label = "" ]
    77 -> 22 [ label = "" ]
    78 -> 23 [ label = "" ]
    64 -> 25 [ label = "" ]
    64 -> 121 [ label = "" ]
    64 -> 77 [ label = "" ]
    64 -> 78 [ label = "" ]
    121 -> 25 [ label = "" ]
    88 -> 26 [ label = "" ]
    88 -> 122 [ label = "" ]
    123 -> 26 [ label = "" ]
    94 -> 26 [ label = "" ]
    94 -> 88 [ label = "" ]
    122 -> 26 [ label = "" ]
    122 -> 123 [ label = "" ]
    79 -> 28 [ label = "" ]
    95 -> 28 [ label = "" ]
    95 -> 79 [ label = "" ]
    91 -> 29 [ label = "" ]
    91 -> 79 [ label = "" ]
    96 -> 29 [ label = "" ]
    96 -> 91 [ label = "" ]
    124 -> 30 [ label = "" ]
    124 -> 87 [ label = "" ]
    125 -> 30 [ label = "" ]
    125 -> 86 [ label = "" ]
    126 -> 30 [ label = "" ]
    126 -> 85 [ label = "" ]
    127 -> 30 [ label = "" ]
    127 -> 84 [ label = "" ]
    128 -> 30 [ label = "" ]
    128 -> 83 [ label = "" ]
    129 -> 30 [ label = "" ]
    129 -> 82 [ label = "" ]
    130 -> 30 [ label = "" ]
    130 -> 81 [ label = "" ]
    57 -> 30 [ label = "" ]
    57 -> 130 [ label = "" ]
    57 -> 129 [ label = "" ]
    57 -> 128 [ label = "" ]
    57 -> 127 [ label = "" ]
    57 -> 126 [ label = "" ]
    57 -> 125 [ label = "" ]
    57 -> 124 [ label = "" ]
    57 -> 80 [ label = "" ]
    56 -> 30 [ label = "" ]
    87 -> 31 [ label = "" ]
    86 -> 31 [ label = "" ]
    85 -> 31 [ label = "" ]
    84 -> 31 [ label = "" ]
    83 -> 31 [ label = "" ]
    82 -> 31 [ label = "" ]
    81 -> 31 [ label = "" ]
    80 -> 31 [ label = "" ]
    80 -> 81 [ label = "" ]
    80 -> 82 [ label = "" ]
    80 -> 83 [ label = "" ]
    80 -> 84 [ label = "" ]
    80 -> 85 [ label = "" ]
    80 -> 86 [ label = "" ]
    80 -> 87 [ label = "" ]
    131 -> 32 [ label = "" ]
    131 -> 26 [ label = "" ]
    132 -> 32 [ label = "" ]
    132 -> 131 [ label = "" ]
    52 -> 32 [ label = "" ]
    52 -> 133 [ label = "" ]
    52 -> 132 [ label = "" ]
    133 -> 32 [ label = "" ]
    134 -> 33 [ label = "" ]
    58 -> 33 [ label = "" ]
    58 -> 134 [ label = "" ]
    135 -> 34 [ label = "" ]
    59 -> 34 [ label = "" ]
    59 -> 135 [ label = "" ]
    102 -> 36 [ label = "" ]
    136 -> 36 [ label = "" ]
    136 -> 29 [ label = "" ]
    137 -> 36 [ label = "" ]
    137 -> 79 [ label = "" ]
    137 -> 91 [ label = "" ]
    138 -> 36 [ label = "" ]
    138 -> 136 [ label = "" ]
    139 -> 36 [ label = "" ]
    97 -> 36 [ label = "" ]
    140 -> 36 [ label = "" ]
    98 -> 36 [ label = "" ]
    98 -> 140 [ label = "" ]
    98 -> 139 [ label = "" ]
    98 -> 138 [ label = "" ]
    98 -> 141 [ label = "" ]
    98 -> 137 [ label = "" ]
    141 -> 36 [ label = "" ]
    142 -> 39 [ label = "" ]
    76 -> 39 [ label = "" ]
    76 -> 142 [ label = "" ]
    143 -> 40 [ label = "" ]
    144 -> 40 [ label = "" ]
    144 -> 92 [ label = "" ]
    71 -> 40 [ label = "" ]
    71 -> 144 [ label = "" ]
    71 -> 143 [ label = "" ]
    103 -> 41 [ label = "" ]
    92 -> 43 [ label = "" ]
    92 -> 101 [ label = "" ]
    101 -> 44 [ label = "" ]
    101 -> 103 [ label = "" ]
    100 -> 46 [ label = "" ]
    100 -> 145 [ label = "" ]
    63 -> 46 [ label = "" ]
    63 -> 145 [ label = "" ]
    99 -> 46 [ label = "" ]
    99 -> 146 [ label = "" ]
    145 -> 46 [ label = "" ]
    146 -> 46 [ label = "" ]
    62 -> 46 [ label = "" ]
    90 -> 48 [ label = "" ]
    89 -> 48 [ label = "" ]
    73 -> 48 [ label = "" ]
    72 -> 48 [ label = "" ]
}

